flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/DataType.java

```diff
@@ -0,0 +1,427 @@
+	 *
+	 * <p>This method exists for explicit declaration of the default behavior or for invalidation of
+	 * a previous call to {@link #notNull()}.
+	 *
+	 * @return a new, reconfigured data type instance
+	 */
+	public abstract DataType andNull();
+
+	/**
+	 * Adds a hint that data should be represented using the given class when entering or leaving
+	 * the table ecosystem.
+	 *
+	 * @return a new, reconfigured data type instance
+	 */
+	public abstract DataType bridgedTo(Class<?> newConversionClass);
+
+	@Override
+	public String toString() {
+		return logicalType.toString();
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		DataType dataType = (DataType) o;
+		return logicalType.equals(dataType.logicalType) &&
+			Objects.equals(conversionClass, dataType.conversionClass);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(logicalType, conversionClass);
+	}
+
+	/**
+	 * This method should catch the most common errors. However, another validation is required in
+	 * deeper layers as we don't know whether the data type is used for input or output declaration.
+	 */
+	private static <C> Class<C> performEarlyClassValidation(
+			LogicalType logicalType,
+			Class<C> candidate) {
+
+		if (candidate != null &&
+				!logicalType.supportsInputConversion(candidate) &&
+				!logicalType.supportsOutputConversion(candidate)) {
+			throw new ValidationException(
+				String.format(
+					"Logical type '%s' does not support a conversion from or to class '%s'.",
+					logicalType.asSummaryString(),
+					candidate.getName()));
+		}
+		return candidate;
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * A data type that does not contain further data types (e.g. {@code INT} or {@code BOOLEAN}).
+	 *
+	 * @see DataTypes for a list of supported data types
+	 */
+	public static final class AtomicDataType extends DataType {
+
+		public AtomicDataType(LogicalType logicalType, @Nullable Class<?> conversionClass) {
+			super(logicalType, conversionClass);
+		}
+
+		public AtomicDataType(LogicalType logicalType) {
+			super(logicalType, null);
+		}
+
+		@Override
+		public DataType notNull() {
+			return new AtomicDataType(
+				logicalType.copy(false),
+				conversionClass);
+		}
+
+		@Override
+		public DataType andNull() {
+			return new AtomicDataType(
+				logicalType.copy(true),
+				conversionClass);
+		}
+
+		@Override
+		public DataType bridgedTo(Class<?> newConversionClass) {
+			return new AtomicDataType(
+				logicalType,
+				Preconditions.checkNotNull(newConversionClass, "New conversion class must not be null."));
+		}
+	}
+
+	/**
+	 * A data type that contains an element type (e.g. {@code ARRAY} or {@code MULTISET}).
+	 *
+	 * @see DataTypes for a list of supported data types
+	 */
+	public static final class ElementDataType extends DataType {
```

Comment:
CollectionDataType? 

It is wired that the `ElementDataType` contains an element type named `elementDataType`.

Reply:
The naming is on purpose as `MapType` is also a collection data type. The naming should not reflect the SQL type families but just serve as a container name.

Reply:
The only concern to me is the class name `ElementDataType` and the member field `elementDataType` is confusing. 

In JDK, `Collection` is a group of elements, and `Map` is not a collection. IMO, `ARRAY` and `MULTISET` are collections of elements with same type.

Reply:
You are right. Java is also a bit confusing in this sense as methods such as `java.util.Collections#emptyMap()` are also put under collections. I will correct that. Also in `org.apache.flink.table.types.logical.LogicalTypeRoot#MAP`. Thanks for the feedback.